electrumx/lib/tx.py

```diff
@@ -306,6 +307,109 @@ def read_tx_and_vsize(self):
+        orig_ser = self.binary[start:self.cursor]
+
+        marker = self._read_byte()
+        flag = self._read_byte()
+
+        # Work around a bug in litecoind v0.21.2 with -rpcserialversion=1 that
+        # returns invalid stripped MW-only transactions with no regular inputs
+        # or outputs, causing this to look like a segwit transaction, but it
+        # really just has zeros for the number of inputs and outputs, which we
+        # incorrectly interpreted as a marker and flag. If what we parsed as
+        # marker and flag are both zero, this is such an invalid transaction.
+        if flag == 0:
+            raise SkipTxDeserialize('invalid MW-only transaction with no regular inputs or outputs')
+
+        start = self.cursor
+        inputs = self._read_inputs()
+        outputs = self._read_outputs()
+        orig_ser += self.binary[start:self.cursor]
+
+        base_size = self.cursor - start
+        
+        # https://github.com/litecoin-project/litecoin/blob/948e6257aec15b52ef68b4e1ee9d73f7c740fae3/src/primitives/transaction.h#L299
+        if flag & 1: # witness flag
+            witness = self._read_witness(len(inputs))
+        else:
+            # A MW tx is allowed to not have the witness bit of the flag byte
+            # set, indicating no witness data encoded for the inputs. Perhaps we
+            # should return a normal Tx here instead, but there is a flag byte
+            # we should probably not just discard.
+            witness = []
+
+        if flag & 8: # MWEB flag
+            # If this transaction is in the main block, not the MW extension
+            # block (MWEB), this should be the HogEx/integration transaction,
+            # which has no mweb tx, just a single zero byte.
+            # https://github.com/litecoin-project/litecoin/blob/bb242e33551157e0db3b70f90f5738f34b82cc51/src/mweb/mweb_node.cpp#L17-L20
+            #
+            # If this is anything but a zero byte, we cannot (yet) deserialize
+            # this transaction since the mwtx at this location is variable
+            # length and we cannot determine the locktime that follows without
+            # accurately parsing it.
+            #
+            # Only in mempool should we encounter transactions with mwtx data,
+            # so for the time being we will simply ignore these transactions.
+            # When such transactions are mined, they will either have the MW
+            # data stripped when they enter the regular block, or if they have
+            # no regular inputs or outputs (MW-only) they will only be in the
+            # MWEB, not the regular block at all.
+            #
+            # Note that by running litecoind with -rpcserialversion=1, the MW
+            # data will be stripped from transactions and this tx flag will be
+            # cleared.
+            if self._read_byte() != 0:
+                raise SkipTxDeserialize('non-null mwtx bytes are not parseable')
```

Comment:
e.g.

`DEBUG:MemPool:skipping tx d9409c0d461c7d7684ba852bba863813203b0c94afeef2717727a09fefb3e141: non-null mwtx bytes are not parseable`

```
$  litecoin-cli getrawtransaction d9409c0d461c7d7684ba852bba863813203b0c94afeef2717727a09fefb3e141 1
{
  "txid": "d9409c0d461c7d7684ba852bba863813203b0c94afeef2717727a09fefb3e141",
  "hash": "d9409c0d461c7d7684ba852bba863813203b0c94afeef2717727a09fefb3e141",
  "version": 2,
  "size": 1530,
  "vsize": 32,
  "weight": 128,
  "locktime": 2351149,
  "vin": [
    {
      "ismweb": true,
      "output_id": "6ebd7e2d0d3eb50c5b83c5d3a09671fe0202fe5be25ce2d95b550a50082fee94"
    },
    {
      "ismweb": true,
      "output_id": "cc33bcf2166d21a827cdf8ed83ba539a4454e6cf278dda9e5d80f1e0a120cb0d"
    }
  ],
  "vout": [
    {
      "ismweb": true,
      "output_id": "f22e230098e0f50d639695edb7f917e8b939349407e0cacee19781b65f441ce7"
    }
  ],
  "vkern": [
    {
      "kernel_id": "41e1b3ef9fa0277771f2eeaf940c3b20133886ba2b85ba84767d1c460d9c40d9",
      "fee": 0.00002840,
      "pegin": 0.00000000,
      "pegout": [
        {
          "value": 0.02102190,
          "scriptPubKey": {
            "asm": "OP_HASH160 4760ada9c0b57700488261a84d24597a799a3a8c OP_EQUAL",
            "hex": "a9144760ada9c0b57700488261a84d24597a799a3a8c87",
            "reqSigs": 1,
            "type": "scripthash",
            "addresses": [
              "MEQa2P4Ud3qASC445xnFFgvf3cQwKobAn8"
            ]
          }
        }
      ]
    }
  ],
  "hex": "0200000000080000010c7568af0f8ec36de2fd6bce090ff9a498dd3ab37bb131a40f493658c14b2a67339506de9b74947476e8f0e1a3daba0f9bb692058e1daacfbc5c7ea2a5955f5f02016ebd7e2d0d3eb50c5b83c5d3a09671fe0202fe5be25ce2d95b550a50082fee94092f199ef2c311bc3cf4670464a408f944cdf54c2c17f50d1699a88cd0f9c5d8d0021df060980c1ca5eb0533912f3a054814de2f33f0643f06d0ee72d2f26b98eaef022fe280879b5cc8f3908beee06d053ac38afc629df6e5b1d93438bfb3b87c91df06b62392cf063ba3e4e7cc892516aa90fc0a5cbd17eac228ed840ea2ebc6c2fd4e01c23fdb99574715bc0e8fa357501e8108e73f2bb17b050f9784cb799a8f0801cc33bcf2166d21a827cdf8ed83ba539a4454e6cf278dda9e5d80f1e0a120cb0d08e69360ceb8cfee3e86f53671bd5d50a68e52f5202f94ba84b88dc40d0b6bd4810227525edb777c32011c8cdbd1703685434634563af82675d3b86ab12062bfcadf02413fa54d5644380f05ff4865b954698409934bef10bf5ad4ec0eb42c7fd9ef6320758f5eb2065237e90fa00848cd507847e3768c23adc45d2cc690a35645a51e9bf5f1fe7457b35c2cd87f7a77ca0f93c948f05a6b963cbbf090605ea6e5cc980109a60f40a9b3b11e61935e87683364e34d6a111988e20f4dc76cfa2ec555890554020acc2706bc2a7f8082c5fec89a069cd56f177d8c994e200012a6057ec4f79574024ecacc186a143ccbb91c120a7fe473ac926f8f2fda521a46d9f1fe6b3eb32bad01023050a296a9377d3274c2c9b0450db07bf64f212453cb28fc1ebb49b900de995fb0a8ebd9cca13aa790596d765f06741060c37e493ad8fd2d7209be677a44fd44189bb36de03cffbe4bce4b42f31c846d3494316745ca1a7bfd3ba4af9a929a8fcfe4f7f094462fcc0000e288d99647ab1063d7fef2a7fbdb360f08b960d956ecf83f0ab3b7ede8787c1d4774c62b8748bdd575ee4bd3abfeeac81e5c30b0423a9e5a457babd7ccbb1d32d4f18a4519aac250def5c31c47116a4409f3d8be615df3418150ff32579015c6eca2db7fd4423f1e667f6e4da6241d2789eb528da9d3739b0ed73c71694ee80b7ca95ed75b58289c20685fcfc6b1df05f25d5792cb61ac9d88b5b03076e2d87e6b0f87a18846b5c31263778ad079ea6c691a651e5e108e3e308dfed3865505addcf109804b291a93018c52812c48b9fa0b5dbb1627573fb274c77f66768bba92a7340a109c803f6ce07dac1cd3380dc89b0b59708c3c92471520b82a5d668c5d105b7dfd4cb2c391553f0d1ab47b586970fe4a26bde9fa412da65c02d0858f8ca58ca330d93b24de07c0ca0fb7723c5d9f0078cce9c39f850ed82ffe6d38bdb263a3523b809d569f98ff2e42fff3a55020b77a84aa71790d7e8bc8802ec61eb5612075e30642d6ae449d9e4b8fe119846ccc283a94325f23e3bb0ecd8d36e65c7e1055ab164275cba38b0c67767142c7b3b8b1691343b70b58af79641029645f2b7f5ca60fbedf39a526fe4971dd2e1cc61ba754db6af861811ae4bf982ff6cc5e91b010e2808e7c225bb24779a3a724bc50ee7656a0288cfe093a7334bdf885a49e998588746fa5edea91ad023607dbaeefc581008f26876795a7ea3e01248ffea04eeeafed2274b14693009b5b122cc73a9df3872a65785872a83c156dec7b3d31712860d37e74e436f3512ea05ce7c058ae444560e6fd794c2441bcfd0462a020d6eee790e27d0201ce22d9b6dd9f3ae1790b5d925407ed61e571bb1e8edbb8198110c88cb04799d92ecc13ee47913ddd436bab4bb413a65277c83f21c469611114bff112b5a88b308acbb587c4d0c291dde7e9a46821520a813dd248aac51d6ec8b711a0b9530b1552acacdc2e3c78b10115951801ffa62e17a9144760ada9c0b57700488261a84d24597a799a3a8c87022b95a1fc1a68ac202f36aaf1a96dfe978d3ab493c51cd930b3fd37de54d783150868917ae5ed73d8337c033eaac6da6e123d17f3d3abb7ac5392ff2b878c41a40b1c5b28c4db9f4c8117482a4bd2548b0934e7e6b018fe8617890be9321dff3830d5bf3175379222837045859bad877b86f22c25b708d928e28552295c5c5e1b5e2de02300"
}
```

With `-rpcserialversion=1`, that mwtx data is would be stripped and the flag is removed, hitting the other `invalid MW-only transaction with no regular inputs or outputs`, but I have not tested in that mode.  But it jives with the stack reported in https://github.com/spesmilo/electrumx/issues/179#issuecomment-1138699874 and my understanding of what `-rpcserialversion=1` does.